### PR TITLE
Upgrade prometheus to v1.2.1

### PR DESCRIPTION
**Pull request by daxmc99**
Prometheus's 1.0 version comes with stronger compatibility guarantees going forward. 
Signed-off-by: Dax McDonald <dax@rancher.com>

### CODE REVIEW

**Comment by adleong** [comment]
Hi @daxmc99!  Is there anything we can do to help with this work?

**Comment by daxmc99** [comment]
Hi @adleong, 
My version bump became a bit more involved since it looks like this will also bump httprouter too  which changes this code https://github.com/linkerd/linkerd2/blob/c9c41e2e8a7573924d8b71365bee5d93d62058c5/controller/tap/handlers.go#L87

See http.router change here: 
https://github.com/julienschmidt/httprouter/blob/08a3b3d20bbeeab9ca4889bcdb65fec0423ef882/router.go#L174 vs
https://github.com/julienschmidt/httprouter/blob/8c199fb6259ffc1af525cc3ad52ee60ba8359669/router.go#L143

I haven't spent too much time on this but it would be nice to know if my handling of the new warnings coming from the Prometheus client is what you all expect to happen. Currently, they are just being logged.

**Comment by adleong** [comment]
I'm not familiar with the new Prometheus API or the semantics of the warnings so I can't say with certainty.  But if these really are just warnings and don't impact that correctness of the returned values, then I think logging them is the correct behavior.

**Comment by daxmc99** [comment]
@adleong  Please take a look

**Comment by daxmc99** [comment]
Hi @adleong, 

I am currently doing some testing on this branch. So far I have not seen any regressions. I did change this from a bump to v1.0.0 to v1.2.1 since there have been some more improvements to the Prometheus client library.  

If would love to see this included in some scale testing since most of my tests have been on my local machine. 

Thanks! 😄

**Comment by ihcsim** [comment]
@daxmc99 Thanks again for this initiative! Apologies for the delay. I haven't had the chance to look further into our scale test setup yet. Meanwhile, if you are interested in taking this a step further, there is the `bin/test-scale` script you can use to get a sense of the before-after-upgrade in-cluster traffic stats. In terms of resource consumption, something like `kubectl -n linkerd top po` will do for now. If you are interested in scale-testing this, we can chat further on slack.

**Comment by ihcsim** [comment]
Here are some observations from my scale test, comparing this branch with edge-19.12.1. The information on the setup can be found at https://github.com/linkerd/linkerd2/issues/3692#issuecomment-564695766

TL;DR: the control plane resource utilization and data plane latencies seem comparable. I am gonna do some more testing with service profiles and traffic splits to make sure nothing regresses there. The lower success rates of the `voting` service were caused by errors simulated by the bots.

## Latency

```
⚡  linkerd version
Client version: edge-19.12.1
Server version: edge-19.12.1

⚡  linkerd stat po -t48h -nlinkerd -owide                              
NAME                                       STATUS   MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
linkerd-controller-6f8976f5db-ct2ps       Running      1/1   100.00%   0.1rps           1ms           2ms           2ms          2          72.3B/s           71.4B/s
linkerd-destination-6f5d6fcdb7-89px6      Running      1/1    98.90%   0.1rps           1ms           9ms          10ms       1797         207.7B/s          207.1B/s
linkerd-grafana-84d8b6f85c-lgtdx          Running      1/1   100.00%   0.1rps           1ms           2ms           2ms          2          84.1B/s           83.3B/s
linkerd-identity-7586d7795b-45bdh         Running      1/1   100.00%   0.1rps           1ms           1ms           1ms       1010          80.6B/s           79.7B/s
linkerd-prometheus-6f99fdb6f5-kc96d       Running      1/1   100.00%   0.0rps           1ms           1ms           1ms          1          11.5B/s           10.7B/s
linkerd-proxy-injector-8546969c57-54n2d   Running      1/1   100.00%   0.1rps           1ms           3ms           3ms          4         105.1B/s          104.2B/s
linkerd-sp-validator-84fd9ff764-58jzl     Running      1/1   100.00%   0.1rps           1ms           1ms           1ms          2          58.2B/s           57.3B/s
linkerd-tap-6c49769c55-4thqt              Running      1/1   100.00%   0.1rps           1ms           2ms           2ms          6         138.0B/s          137.2B/s
linkerd-web-7494d77777-5lpsr              Running      1/1   100.00%   0.1rps           1ms           1ms           1ms          2          58.2B/s           57.3B/s

⚡  linkerd stat po -t48h -owide -nlinkerd-emoji-20
NAME                         STATUS   MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
emoji-6bb7854db4-5nk4t      Running      1/1   100.00%   0.4rps           1ms           1ms           1ms          6         533.9B/s          536.9B/s
emoji-6bb7854db4-682b7      Running      1/1   100.00%   0.2rps           1ms           2ms           3ms          6         207.3B/s          208.3B/s
emoji-6bb7854db4-nzbb6      Running      1/1   100.00%   0.4rps           1ms           3ms           4ms          6         531.9B/s          534.9B/s
emoji-6bb7854db4-sx9qz      Running      1/1   100.00%   0.3rps           1ms           6ms           9ms          6         355.6B/s          357.8B/s
emoji-6bb7854db4-zwbl4      Running      1/1   100.00%   0.5rps           1ms           1ms           1ms          6         603.6B/s          606.8B/s
vote-bot-5b464848c4-2gccv   Running      1/1         -        -             -             -             -          -                -                 -
vote-bot-5b464848c4-8cnmb   Running      1/1         -        -             -             -             -          -                -                 -
vote-bot-5b464848c4-cqbcj   Running      1/1         -        -             -             -             -          -                -                 -
vote-bot-5b464848c4-sjzvx   Running      1/1         -        -             -             -             -          -                -                 -
vote-bot-5b464848c4-zzsmv   Running      1/1         -        -             -             -             -          -                -                 -
voting-869b78c895-4tbj4     Running      1/1    84.20%   0.2rps           1ms           7ms           9ms          6          26.0B/s           27.6B/s
voting-869b78c895-6f5zt     Running      1/1    84.06%   0.2rps           1ms          24ms          29ms          6          36.8B/s           39.0B/s
voting-869b78c895-9tmtm     Running      1/1    84.27%   0.2rps           1ms           3ms           3ms          6          36.2B/s           38.3B/s
voting-869b78c895-l5x47     Running      1/1    84.37%   0.2rps           1ms           1ms           2ms          6          28.8B/s           30.5B/s
voting-869b78c895-qnkgh     Running      1/1    84.04%   0.2rps           1ms           5ms           9ms          6          32.2B/s           34.2B/s
web-7f7b6486f4-4pzb6        Running      1/1    91.98%   0.5rps           4ms           9ms          74ms          8        1139.0B/s         1195.7B/s
web-7f7b6486f4-fwbhl        Running      1/1    92.20%   0.4rps           7ms          67ms          93ms          8         867.8B/s          911.0B/s
web-7f7b6486f4-jxdb6        Running      1/1    91.98%   0.5rps           4ms          10ms          90ms          7        1140.1B/s         1196.5B/s
web-7f7b6486f4-k79cw        Running      1/1    92.30%   0.4rps           5ms          16ms          19ms          7         943.4B/s          990.4B/s
web-7f7b6486f4-mb4cd        Running      1/1    92.05%   0.5rps           5ms          28ms          47ms          8        1225.3B/s         1285.8B/s

```

```
⚡  linkerd version                                                     
Client version: edge-19.12.1
Server version: git-8ef3ca6c

⚡  linkerd stat po -t48h -nlinkerd -owide 
NAME                                       STATUS   MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
linkerd-controller-85f6ffccc7-phjx7       Running      1/1   100.00%   0.1rps           1ms           2ms           2ms          2          71.9B/s           71.1B/s
linkerd-destination-6d8f49cf96-5b66k      Running      1/1    99.41%   0.1rps           1ms           9ms          10ms       1316         195.1B/s          194.4B/s
linkerd-grafana-6c8d774db-5g9nd           Running      1/1   100.00%   0.1rps           1ms           4ms           4ms          2          84.0B/s           83.1B/s
linkerd-identity-57c89ffdff-sfw94         Running      1/1   100.00%   0.1rps           1ms           2ms           2ms       1010          71.4B/s           70.5B/s
linkerd-prometheus-64db8bd7dc-gbqwz       Running      1/1   100.00%   0.0rps           1ms           1ms           1ms          9          11.6B/s           10.7B/s
linkerd-proxy-injector-578b59bb6b-dcsrz   Running      1/1   100.00%   0.1rps           1ms           4ms           4ms          4          97.3B/s           96.4B/s
linkerd-sp-validator-8664bb899-mzwpl      Running      1/1   100.00%   0.1rps           1ms           1ms           1ms          2          57.9B/s           57.1B/s
linkerd-tap-7479c55f96-jp8lx              Running      1/1   100.00%   0.1rps           1ms           2ms           2ms          6         133.3B/s          132.5B/s
linkerd-web-58dd6d467d-vcjt9              Running      1/1   100.00%   0.1rps           1ms           2ms           2ms          2          57.9B/s           57.1B/s

⚡  linkerd stat po -t48h -owide -nlinkerd-emoji-44
NAME                         STATUS   MESHED   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TCP_CONN   READ_BYTES/SEC   WRITE_BYTES/SEC
emoji-6bb7854db4-4bklf      Running      1/1   100.00%   0.3rps           1ms           1ms           1ms          6         379.1B/s          381.5B/s
emoji-6bb7854db4-jlgbj      Running      1/1   100.00%   0.5rps           1ms           1ms           1ms          6         617.6B/s          620.8B/s
emoji-6bb7854db4-p7b9q      Running      1/1   100.00%   0.4rps           1ms           1ms           2ms          6         512.8B/s          515.9B/s
emoji-6bb7854db4-r7f42      Running      1/1   100.00%   0.4rps           1ms           1ms           1ms          6         540.0B/s          543.3B/s
emoji-6bb7854db4-rdczj      Running      1/1   100.00%   0.5rps           1ms           4ms           5ms          6         647.1B/s          650.5B/s
vote-bot-5b464848c4-6bstg   Running      1/1         -        -             -             -             -          -                -                 -
vote-bot-5b464848c4-9fp9q   Running      1/1         -        -             -             -             -          -                -                 -
vote-bot-5b464848c4-dmkxj   Running      1/1         -        -             -             -             -          -                -                 -
vote-bot-5b464848c4-mwr7m   Running      1/1         -        -             -             -             -          -                -                 -
vote-bot-5b464848c4-sxwkq   Running      1/1         -        -             -             -             -          -                -                 -
voting-869b78c895-26lwl     Running      1/1    84.70%   0.2rps           1ms          24ms          29ms          6          30.6B/s           32.6B/s
voting-869b78c895-8zgwz     Running      1/1    84.38%   0.3rps           1ms           1ms           1ms          6          40.8B/s           43.2B/s
voting-869b78c895-cb7d7     Running      1/1    84.13%   0.2rps           1ms           3ms           3ms          6          30.1B/s           32.0B/s
voting-869b78c895-n754c     Running      1/1    84.29%   0.2rps           1ms           1ms           1ms          6          34.9B/s           37.1B/s
voting-869b78c895-vw2d2     Running      1/1    83.90%   0.2rps           1ms           1ms           1ms          6          36.3B/s           38.6B/s
web-7f7b6486f4-6p9q9        Running      1/1    92.22%   0.5rps           4ms          12ms          18ms          8        1103.5B/s         1158.0B/s
web-7f7b6486f4-bgtcq        Running      1/1    92.26%   0.3rps           4ms          16ms          27ms          7         640.5B/s          672.1B/s
web-7f7b6486f4-bn7vr        Running      1/1    92.27%   0.5rps           4ms           9ms          10ms          7        1236.5B/s         1298.1B/s
web-7f7b6486f4-clqps        Running      1/1    92.09%   0.4rps           5ms          10ms          10ms          7         956.3B/s         1004.2B/s
web-7f7b6486f4-fp6zp        Running      1/1    91.99%   0.4rps           5ms          16ms          19ms          7        1023.3B/s         1074.4B/s
```

**Comment by ihcsim** [comment]
## CPU Utilization 

This is expressed as % of the defined limits.

CPU utilization of #3541 control plane over 6 hours:
dev-cpu-utilization

CPU utilization of edge-19.12.1 control plane over 6 hours:
edge-cpu-utilization


## Memory Usage

Memory usage of #3541 control plane over 6 hours:
dev-memory-usage

Memory usage of edge-19.12.1 control plane over 6 hours:
edge-memory-usage

**Comment by ihcsim** [comment]
TL;DR: No significant differences in the metrics and timeseries counts between edge-19.12.1 and this branch.

## Metrics Counts

This count is grouped by the Prometheus jobs.

Metrics count for edge-19.12.1 after 6 hours:
edge-top10-metrics

Metrics count for #3541 after 6 hours:
dev-top10-metrics

## Proxy's Time Series Counts

Proxy's time series count for edge-19.12.1 after 6 hours:
edge-top10-proxy-metrics

Proxy's time series count for #3541 after 6 hours:
dev-top10-proxy-metrics

**Comment by daxmc99** [comment]
@ihcsim Great to see this and thank you for taking the time to scale test this. 
Glad to see that everything looks ok.

**Comment by ihcsim** [comment]
Both service profiles and traffic splits continue to work on this branch :+1: :heart:. 

```
⚡  linkerd -n emojivoto routes deploy                                 
==> deployment/emoji <==
ROUTE                         SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
[DEFAULT]                   emoji-svc         -        -             -             -             -
emojivoto.v1.EmojiService   emoji-svc   100.00%   2.0rps           1ms           1ms           1ms

==> deployment/vote-bot <==
ROUTE   SERVICE   SUCCESS   RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99

==> deployment/voting <==
ROUTE                           SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
[DEFAULT]                    voting-svc         -        -             -             -             -
emojivoto.v1.VotingService   voting-svc    91.67%   1.0rps           1ms           1ms           1ms

==> deployment/web <==
ROUTE       SERVICE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
/api/list   web-svc   100.00%   1.0rps           3ms           4ms           4ms
/api/vote   web-svc         -        -             -             -             -
[DEFAULT]   web-svc    91.67%   1.0rps           4ms           4ms           4ms
```

image

**Comment by ihcsim** [comment]
All the integration tests have passed:
```
=== PASS: all tests passed
```

**Comment by daxmc99** [comment]
Awesome!! 🎆

**Comment by ihcsim** [comment]
@daxmc99 Thanks again for this initiative. Please fill up this [form](https://docs.google.com/forms/d/e/1FAIpQLSdXA7kAl6oyUQ4DFRjiDc2biWlvkhNirrMwsXaQ8V3Uh8Yj5A/viewform), so that we can send you some awesome Linkerd swags.
